### DISCUSSION TITLE
how to implement transaction?

### DISCUSSION THREAD

**Discussion by xu4wang** [directus/directus#17257] [Archive]
For example, I have several steps updating different items, I want all those updates all successful, otherwise none of them should be executed. 

I was thinking about "transaction... commit ... rollback" but I am not sure how to implement using directus API/SDK?

Any suggestion is welcome!

**Comment by paescuj** [comment]
This is (currently) not possible via the API, but you could create a [Custom API Hooks extension](https://docs.directus.io/extensions/hooks.html#custom-api-hooks) to do all those updates. With the [filter](https://docs.directus.io/extensions/hooks.html#filter) event you've access to the current transaction (`database` property).

**Reply by quyendaoqt707** [reply]
Hi maintainer,

I am struggling with the transaction in the custom Filter hook. According to the documentation you mentioned, the `context.database` should use the current transaction (of the current request?). However, according to the `Knex.js` logs, the next delete handler (by Directus) is using a different transaction. This means that if the delete action by the Directus handler fails for any reason, the filter hook action will not be rolled back. 

Is this behavior intended, a bug, or am I missing something?

Thank you!

Here is my code, and the attached image is the corresponding log:
```ts
// Hook event: 'comment.items.delete'
const filterHandlerForDelete = (hookContext: HookExtensionContext) => (
	async  (input: any , meta: MetaActionRegister, context: EventContext)  =>{

	// Get the comment that will be deleted using the current transaction
	const commentToDelete = await context.database
		.from('community_comment')
		.where('id', input[0]) // input is an array containing the ID
		.first();

	console.log('⏭️ Comment to be deleted:', commentToDelete);


	// Update the post's comment_count with context
	await context.database.from('post')
			.where('id', commentToDelete.post_id)
			.decrement('comment_count', 1)

	return input
})
```
![image](https://github.com/user-attachments/assets/80b62550-7f56-41fb-9f12-f9ad73e9bc64)

**Reply by rijkvanzanten** [reply]
The filter hook is fired before the deletion is executed and is intended for hooks to be able to cancel out of the deletion by throwing an error. It is indeed currently not part of the same transaction.

**Reply by SadmanYasar** [reply]
Would it be as simple as adding a button in the flow where we decide if it's a transaction or not in action types? Or have to stick to using hooks for complex transactions?

**Reply by SadmanYasar** [reply]
@rijkvanzanten would adding Supabase on top of Directus then writing db functions with transactions work with Directus API? I am thinking about this since Supabase already provides an SQL editor so adding Directus at the API level and Supabase specifically for transactions. Also this shouldn't affect any Flows in Directus given that these are action and not filter.
